### PR TITLE
[Feature] Add DOLG neck for image retrieval

### PR DESCRIPTION
**Pull request by zzc98**
## Motivation

The implementation of DOLG, which is a kind of neck network in the retrieval task.  More details about DOLG can be found in [DOLG: Single-Stage Image Retrieval with Deep Orthogonal Fusion of Local and Global Features](https://arxiv.org/abs/2108.02927).

## Use cases

```python
model = dict(
    type='ImageClassifier',
    backbone=dict(
        type='ResNet',
        depth=101,
        num_stages=4,
        out_indices=(3,),
        style='pytorch'
    ),
    neck=dict(type='DolgNet',
               local_dim=1024,
               global_dim=2048,
               hidden_dim=1024,
               dilation_rates=(3, 6, 9)
    ),
    head=dict(
        type='LinearClsHead',
        num_classes=1000,
        in_channels=2048,
        loss=dict(type='CrossEntropyLoss', loss_weight=1.0),
        topk=(1, 5),
    ),
)
```

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcls/models/necks/dolg.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            dilation_rates,
+            int), ('``dilation_rates`` must be a sequence of int.')
+        assert out_channel % 2 == 0
+
+        # In the original implementation, here hidden-dim was set to a
+        # fixed 512 as it was designed for resnet101. We set it here to
+        # half of out_c for the sake of generality, which is consistent
+        #  when using resnet101.
+        hidden_channel = out_channel // 2
+        self.dilated_conv_list = nn.ModuleList([
```
Comment (okotaku) [inline mmcls/models/necks/dolg.py:42]: https://github.com/feymanpriv/DOLG/blob/63b117d76f660db9617c2081ec771e3cbdf4bcc8/core/model/dolg_model.py#L143
`dilation=1` conv may be needed.

File: mmcls/models/necks/dolg.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+        # convolution after concat for smoothing
+        num = len(dilation_rates) + 1
+        self.conv_after = nn.Sequential(
+            nn.Conv2d(hidden_channel * num, out_channel, kernel_size=(1, 1)),
+            nn.ReLU())
+
+    def forward(self, x):
+        h, w = x.shape[2], x.shape[3]
+        local_feat_list = [
```
Comment (okotaku) [inline mmcls/models/necks/dolg.py:65]: There are some aspects that differ from some unofficial implementations, including this section. Is the accuracy reproduced?

https://github.com/dongkyuk/DOLG-pytorch/blob/main/model/dolg.py
https://github.com/ChristofHenkel/kaggle-landmark-2021-1st-place/blob/main/models/ch_mdl_dolg_efficientnet.py

Comment (Ezra-Yu) [reply mmcls/models/necks/dolg.py:65]: the same with https://github.com/dongkyuk/DOLG-pytorch/blob/9726243e8b150e429674f1f41ddacd79d215a2ec/model/dolg.py#L28 `nn.Upsample` just use `F.interpolate(mode='bilinear',)` 
 
I think it would be better to use the official implementation here.

The precision here needs to be supported by the dataset, so we won't reproduce the precision for now

File: mmcls/models/necks/dolg.py (hunk 1 above)
Comment (Ezra-Yu) [reply mmcls/models/necks/dolg.py:42]: yes, and the kernel_size should be 1 too.

Comment (Ezra-Yu) [reply mmcls/models/necks/dolg.py:48]: https://github.com/feymanpriv/DOLG/blob/63b117d76f660db9617c2081ec771e3cbdf4bcc8/core/model/dolg_model.py#L143 is for `self.gap_conv`, its dilation is 1.

self.dilated_conv_lists' `dilation`  are `rate` refer to https://github.com/feymanpriv/DOLG/blob/63b117d76f660db9617c2081ec771e3cbdf4bcc8/core/model/dolg_model.py#L147.

To avoid confusion in reading the code, I will put them in reverse order

**Review by okotaku — APPROVED** [review]
I understand. LGTM

**Review by Ezra-Yu — APPROVED** [review]
LGTM.
